token/program-2022/src/extension/mod.rs

```diff
@@ -2122,4 +2162,89 @@ mod test {
             .unwrap();
         assert_eq!(new_len, current_len);
     }
+
+    /// Test helper for mimicking the data layout an on-chain `AccountInfo`,
+    /// which permits "reallocs" as the Solana runtime does it
+    struct SolanaAccountData {
+        data: Vec<u8>,
+        lamports: u64,
+        owner: Pubkey,
+    }
+    impl SolanaAccountData {
+        /// Create a new fake solana account data. The underlying vector is
+        /// overallocated to mimic the runtime
+        fn new(account_data: &[u8]) -> Self {
+            let mut data = vec![];
+            data.extend_from_slice(&(account_data.len() as u64).to_le_bytes());
+            data.extend_from_slice(account_data);
+            data.extend_from_slice(&[0; MAX_PERMITTED_DATA_INCREASE]);
+            Self {
+                data,
+                lamports: 10,
+                owner: Pubkey::new_unique(),
+            }
+        }
+
+        /// Data lops off the first 8 bytes, since those store the size of the
+        /// account for the Solana runtime
+        fn data(&self) -> &[u8] {
+            let start = size_of::<u64>();
+            let len = self.len();
+            &self.data[start..start + len]
+        }
+
+        /// Gets the runtime length of the account data
+        fn len(&self) -> usize {
+            self.data
+                .get(..size_of::<u64>())
+                .and_then(|slice| slice.try_into().ok())
+                .map(u64::from_le_bytes)
+                .unwrap() as usize
+        }
+    }
+    impl GetAccount for SolanaAccountData {
+        fn get(&mut self) -> (&mut u64, &mut [u8], &Pubkey, bool, Epoch) {
+            // need to pull out the data here to avoid a double-mutable borrow
+            let start = size_of::<u64>();
+            let len = self.len();
+            (
+                &mut self.lamports,
+                &mut self.data[start..start + len],
+                &self.owner,
+                false,
+                Epoch::default(),
+            )
+        }
+    }
+    #[test]
+    fn alloc_new_tlv_in_account_info() {
+        const VALUE_LEN: usize = 10;
+        let base_account_size = Mint::LEN;
+        let mut buffer = vec![0; base_account_size];
+        let mut state = StateWithExtensionsMut::<Mint>::unpack_uninitialized(&mut buffer).unwrap();
```

Comment:
It would be nice if this test had a sized extension also, to ensure alloc_and_serialize plays nicely with existing extensions.

Reply:
This is intentional -- it caught an implementation bug where I forgot to set the account type when going from the base 82 bytes to the extended 165+. The realloc test starts with a 165+ account that already has an extension, so that case is covered there.

Either way, I can add a separate alloc test for that and a comment explaining why this test matters too

Reply:
Perfect :)